Move DL items from gfx to layout and implement corner clipping

pyfisch opened the pull request:
Implement corner clipping.
Remove PixelFormat from WebrenderImageInfo.
Use WebRender text shadow.
Remove MallocSizeOf and Deserialize for DL items.

Closes #19649, closes #19680, closes #19802

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/20420)
<!-- Reviewable:end -->

pyfisch committed:
Move DL items from gfx to layout

Implement corner clipping.
Remove PixelFormat from WebrenderImageInfo.
Use WebRender text shadow.
Remove MallocSizeOf and Deserialize for DL items.

Closes #19649, #19680, #19802

highfive commented:
Heads up! This PR modifies the following files:
 * @emilio: components/layout/block.rs, components/layout/animation.rs, components/layout/display_list/background.rs, components/layout/display_list/items.rs, components/layout/opaque_node.rs and 10 more

pyfisch commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 26ace4cdcdd0fe80d240216353b0388dd17d53d7 with merge 50b6e6365bea6236f39a8479a9cd1c71cda1bb3f...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11500)

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/44.

pyfisch commented:
http://build.servo.org/builders/linux-rel-wpt/builds/8339/steps/shell__4/logs/filtered-wpt-errorsummary.log

pyfisch commented:
This is now ready for review.

The two failing tests are some weird bug that only occurs for elements with `direction: rtl` in certain conditions. It appears that the WritingMode is empty for these elements while the border_padding has direction RTL. This causes an assertion in debug mode to fail.

pyfisch commented:
@Manishearth: You created the "build_display_list_for_background_if_applicable_with_background" method. I guess based on the name it should display the provided background instead of the element background. But in only uses the provided background for color and [not for images and gradients](https://github.com/servo/servo/blob/36a41722b6b93be6e500b4fc0debfa9b5d2ac7bf/components/layout/display_list/builder.rs#L874). (It does not suffice to remove the line because the different images again access the background via `style: &ComputedValues`.

Manishearth commented:
Ah, oops. It should be using the same background. You're right that that's not enough; fixing the rest is a part of the fix for https://github.com/servo/servo/issues/20093 (which I haven't had time to get to)

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#44.

pyfisch commented:
@bors-servo try

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#44.

pyfisch commented:
@bors-servo try

pyfisch commented:
r? @emilio

This PR moves the display list items from the gfx crate to layout. They are only ever used in layout and moving them there will hopefully decrease compile when modifying these items as gfx stays untouched. This also adds corner clipping for backgrounds along with a test. 

Previously this used the same code to clip the background color and the background images but I reverted it because of two failing right-to-left tests.

As you see I tried to start bors but for some reason the bot is not replying?

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#44.

emilio commented on components/gfx/platform/macos/font_context.rs:
```diff
@@ -2,7 +2,9 @@
-#[derive(Clone, Debug, MallocSizeOf)]
+use malloc_size_of::{MallocSizeOf, MallocSizeOfOps};
+
+#[derive(Clone, Debug)]
```
I don't understand this, why is `#[derive()]` not usable anymore?

pyfisch replied:
`#[derive()]` is still usable but it is only once used in the whole crate and only on MacOS. (I didn't first notice because on Linux it built fine.) Should I add the dependency back and use `#[derive()]` again? Btw are there platform specific dependencies?

emilio commented on components/layout/display_list/items.rs:
```diff
@@ -74,15 +71,18 @@ impl ClippingAndScrolling {
+    pub fn new(
+        scrolling: ClipScrollNodeIndex,
+        clipping: ClipScrollNodeIndex,
+    ) -> ClippingAndScrolling {
```
nit: This can use `Self` and fit into the same line I think.

emilio commented on components/layout/display_list/builder.rs:
```diff
@@ -967,13 +980,21 @@ impl FragmentDisplayListBuilding for Fragment {
             index,
         );
 
+        let previous_clipping_and_scrolling = state.current_clipping_and_scrolling;
```
This looks somewhat easy to mess up returning before we get to the end of the function or what not. Could we use some sort of:

````
state.with_clipping_and_scrolling(|state| {
    do_foo(state);
})
```

that takes care of setting and restoring the state?

pyfisch replied:
I think using a closure is a great idea. But I would want to set the state manually as in many case only part of the time a new clipping is actually added.

```rust
state.with_clipping_and_scrolling(|state| {
    do_foo(state);
    if something {
        state.current_clipping_and_scrolling = new_clip();
    }
    do_bar(state);
})
```

emilio replied:
Ok, that makes sense :)

pyfisch replied:
I have added `with_clipping_and_scrolling` in a new commit.

emilio commented on components/layout/display_list/items.rs:
```diff
@@ -986,19 +995,16 @@ impl WebRenderImageInfo {
-        self.m21 == _0 && self.m22 == _1 && self.m23 == _0 && self.m24 == _0 &&
-        self.m31 == _0 && self.m32 == _0 && self.m33 == _1 && self.m34 == _0 &&
-        self.m44 == _1
+        let (_0, _1) = (0.0, 1.0);
```
This looks unrelated, and all the unrelated changes makes this somewhat hard to review.

pyfisch replied:
Sorry about making review really hard by rustfmting everything.

emilio commented on components/layout/display_list/items.rs:
```diff
@@ -935,14 +943,17 @@ impl fmt::Debug for DisplayItem {
-                            solid_color.color.b,
-                            solid_color.color.a),
+                DisplayItem::SolidColor(ref solid_color) => format!(
+                    "SolidColor rgba({}, {}, {}, {})",
```
Same for these, I'm fine with formatting changes, but they should be a different commit.

emilio commented on components/layout/display_list/items.rs:
```diff
@@ -550,13 +550,14 @@ impl ClippingRegion {
-                ComplexClipRegion {
+            complex: self.complex
+                .iter()
+                .map(|complex| ComplexClipRegion {
```
I think this is uglier btw, but I assume this has been done with rustfmt... Oh well.

emilio commented on components/layout/display_list/builder.rs:
```diff
@@ -1066,12 +1088,20 @@ impl FragmentDisplayListBuilding for Fragment {
             index,
         );
 
+        let previous_clipping_and_scrolling = state.current_clipping_and_scrolling;
```
Here we could use that helper too, ideally moving the inner block to another function or what not.

pyfisch replied:
Yes this pattern is used two times already in the file. This PR adds two more I think. The helper could handle those all.

emilio reviewed (commented):
Generally it looks reasonable, though this is hard to review with all the unrelated formatted and behavior changes that have been introduced here. Any chance you could split this out into different commits?

@mrobinson should at least look at the clipping bits since he's more familiar than me on that.

Thank you for working on this!

pyfisch commented:
> Any chance you could split this out into different commits?

There were different commits (plus some additional changes made later) but the branch can't be automatically merged with master. I think I combined all those commits before I created this PR so I didn't need to manually merge many times and build and test each time.

These are the commits: https://github.com/pyfisch/servo/tree/abcd

Do these commits help with review?
I can try to create a new PR based on these commits. Which commits can be combined and which ones should stay separate?

emilio commented:
Thanks! That tree was really helpful :)

Ok, r=me on the moving the display items to layout. I can review the clipping changes tomorrow morning, or @mrobinson can too. Probably the PR title should also be adjusted.

pyfisch committed:
Place item clipping_and_scrolling inside closure

A utility function assures that if the clipping and scrolling are changed
for a display item the old value is later restored.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#44.

emilio reviewed (approved):
r=me, with comments addressed. Thank you!

pyfisch commented on components/layout/display_list/builder.rs:
```diff
@@ -858,25 +864,24 @@ impl FragmentDisplayListBuilding for Fragment {
-            base: base,
-            color: background_color.to_layout(),
-        })));
+        state.with_clipping_and_scrolling(|state| {
```
Maybe rename with_clipping_and_scrolling as with_element_clip?
I think this is what it does. It allows an element to have its own clip without affecting other parts of the layout

emilio commented:
Thank you!

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit aada975 has been approved by `emilio`

<!-- @bors-servo r=emilio aada975dea2a31cb6879be63d91be1299d9f2147 -->

bors-servo commented:
:hourglass: Testing commit aada975dea2a31cb6879be63d91be1299d9f2147 with merge 54d0ec008dd8e550ba7825d1818b26ab803103e8...

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/53.

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4242)

pyfisch commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/11386), [arm32](http://build.servo.org/builders/arm32/builds/11619), [arm64](http://build.servo.org/builders/arm64/builds/11663), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11699), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8595), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8598), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4145), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4172), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8419), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8481), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4185), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7091) are reusable. Rebuilding only [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11747), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4242)...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11755)

pyfisch commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/11386), [arm32](http://build.servo.org/builders/arm32/builds/11619), [arm64](http://build.servo.org/builders/arm64/builds/11663), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11699), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8595), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8598), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4145), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4172), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8419), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8481), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4185), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7091) are reusable. Rebuilding only [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11755), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4246)...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4246)

nox commented:
@bors-servo retry

pyfisch commented:
@bors-servo retry

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/54.

nox commented:
Why do you keep closing and reopening it?

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/55.

pyfisch commented:
@nox because bors did not react to your comment half an hour ago. Closing and opening worked last time when the bot ignored the PR and emilo brought it back to attention.

pyfisch commented:
@bors-servo try

nox commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit aada975dea2a31cb6879be63d91be1299d9f2147 with merge 203ef3813a7c37c8194964f8b76780cd4082e7de...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8514)

bors-servo commented:
:boom: Test timed out

pyfisch commented:
@emilio, @nox I don't know what "test timed out" means. This PR didn't pass CI now so many times (for reasons probably outside my code) that I give up trying.

Can you merge the PR for me? Tell me if the code needs changes.

nox commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit aada975dea2a31cb6879be63d91be1299d9f2147 with merge 7f1311d0f5f5aa845999903d283771e5619b9110...

CYBAI commented:
```
{"status": "FAIL", "group": "default", "message": "ReferrerPolicyTestCase is not defined", "stack": "@http://web-platform.test:8000/referrer-policy/origin-when-cross-origin/http-rp/cross-origin/http-https/iframe-tag/cross-origin.no-redirect.http.html:23:7\n", "subtest": "Referrer-Policy: Referrer Policy is set to 'origin-when-cross-origin'", "test": "/referrer-policy/origin-when-cross-origin/http-rp/cross-origin/http-https/iframe-tag/cross-origin.no-redirect.http.html", "line": 179025, "action": "test_result", "expected": "PASS"}
{"status": "OK", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/referrer-policy/origin-when-cross-origin/http-rp/cross-origin/http-https/iframe-tag/cross-origin.no-redirect.http.html", "line": 179026, "action": "test_result", "expected": "TIMEOUT"}
```

emilio commented:
@bors-servo r+

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.
- There's another pull request that is currently being tested, blocking this pull request: #20678

bors-servo commented:
:pushpin: Commit aada975 has been approved by `emilio`

<!-- @bors-servo r=emilio aada975dea2a31cb6879be63d91be1299d9f2147 -->

emilio commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit aada975dea2a31cb6879be63d91be1299d9f2147 with merge 0ff6f32d7d4eb597a1ea9ea7480c0371b79386b4...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/11432), [arm32](http://build.servo.org/builders/arm32/builds/11663), [arm64](http://build.servo.org/builders/arm64/builds/11709), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11745), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8642), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8646), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11795), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4195), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4222), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8465), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8528), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4290), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4228), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7138)
Approved by: emilio
Pushing 0ff6f32d7d4eb597a1ea9ea7480c0371b79386b4 to master...
